cmd/sink/expr/cel.go

```diff
@@ -0,0 +1,312 @@
+	if err != nil {
+		return err
+	}
+	resources := []v1alpha1.KubeArchiveConfigResource{}
+	err = yaml.Unmarshal(resourcesBytes, &resources)
+	if err != nil {
+		return err
+	}
+	var errList []error
+	f.Lock()
+	defer f.Unlock()
+	for _, resource := range resources {
+		gvk := schema.FromAPIVersionAndKind(resource.Selector.APIVersion, resource.Selector.Kind)
+		namespaceGvk := NamespaceGroupVersionKind(namespace, gvk)
+		archiveExpr, err := CreateCelExprOr(resource.ArchiveWhen, globalArchive[gvk.String()])
+		if err != nil {
+			errList = append(errList, err)
+		} else {
+			f.archive[namespaceGvk] = *archiveExpr
+		}
+		deleteExpr, err := CreateCelExprOr(resource.DeleteWhen, globalDelete[gvk.String()])
+		if err != nil {
+			errList = append(errList, err)
+		} else {
+			f.delete[namespaceGvk] = *deleteExpr
+		}
+	}
+	return errors.Join(errList...)
+}
+
+// Update the archive and delete filters when a ConfigMap file changes
+func (f *Filters) Update(watcher *fsnotify.Watcher) {
+	logger := log.New(os.Stderr, "", log.LstdFlags|log.Lmicroseconds|log.LUTC)
+	for {
+		select {
+		case event, ok := <-watcher.Events:
+			if !ok { // watcher.Close() was called
+				return
+			}
+			switch event.Op {
+			case fsnotify.Create:
+				logger.Printf("File %s was created. Updating filters accordingly\n", event.Name)
+				fileName := filepath.Base(event.Name)
+				dir := filepath.Dir(event.Name)
+				filePaths, err := files.DirectoryFiles(dir)
+				if err != nil {
+					logger.Printf("Could not get all files in directory %s: %s\n", dir, err)
+					continue
+				}
+				globalArchive, globalDelete, err := getGlobalCelExprs(filePaths)
+				if fileName == globalKey {
+					logger.Println("Updating global filters")
+					err := f.changeGlobalFilters(filePaths, globalArchive, globalDelete)
+					if err != nil {
+						logger.Println("Problem updating global filters\n", err)
+					}
+					logger.Println("Successfully updated global keys")
+					continue
+				}
+				logger.Println("Updating filters for namespace:", fileName)
+				err = f.createFilters(fileName, event.Name, globalArchive, globalDelete)
+				if err != nil {
+					logger.Printf("Problem updating filters for namespace %s: %s\n", fileName, err)
+				}
+				logger.Println("Successfully updated filters for namespace:", fileName)
+			}
+		// TODO: handle other event.OP cases
+		}
+	}
+}
+
+// Returns whether obj needs to be archived. Obj needs to be archived if any of the cel programs in archive return true
+// or if obj needs to be deleted. If obj is nil, it returns false
+func (f *Filters) MustArchive(ctx context.Context, obj *unstructured.Unstructured) bool {
+	if obj == nil {
+		return false
+	}
+	f.RLock()
+	defer f.RUnlock()
+	mustArchive := f.mustDelete(ctx, obj)
+	if mustArchive {
+		return mustArchive
+	}
+	gvk := NamespaceGroupVersionKindFromObject(obj)
+	program, exists := f.archive[gvk]
+	if exists {
+		mustArchive = executeCel(ctx, program, obj)
+	}
+	return mustArchive
```

Comment:
I think you're right. I didn't think to do something like this because the trade off being made is compactness of the code for ease of following the boolean logic. If you don't think this is too complicated I can update this one too

Reply:
I do not. To me it's much cleaner.

Reply:
done